### DISCUSSION TITLE
BlazorServer : ApplicationContext issue on IIS Hosting

### DISCUSSION THREAD

**Discussion by sylum** [MarimerLLC/csla#2714] [Questions] [answered]
Version and Platform
CSLA : 5.4.2
BLAZOR SERVER
VS2019
WINDOW10
Application : ProjectTracker

Hi Folk,

I'm facing an issue with the Blazor Server application that can be produced with the Project Tracker solution.
The BlazorServer application works fine when running on IISExpress but failed when deployed on IIS.
Actually, it's little more complicated cause the problem only appears on some PCs.
For information, I have 2  PCs => both WIN10 (same updates), same SDK DotNetCore, same VS2019 and same IIS options.
- The First one make it run the BlazorServer application perfectly.
- The second one give me the following error message :
	

> "Unable to cast object of type 'Csla.Security.UnauthenticatedPrincipal' to type 'System.Security.Claims.ClaimsPrincipal"

- I have been also able to produce the issue on a VM as well.

Using the PC (the one with the issue) and after targeting IIS with VS2019, I have set a breakpoint on the page Index (see code below) and use the Immediate window to verify that the ApplicationContext.User property is correct.
The result is the ApplicationContext.User is well set to **System.Security.Claims.ClaimsPrincipal** and the first call to appserver works (returning data) but when the the blazor page is reloaded the ApplicationContext.User become
**Csla.Security.UnauthenticatedPrincipal** and raise the exception.

```
@code {
    protected override async Task OnParametersSetAsync()
    {
        try
        {
            var user = Csla.ApplicationContext.User;
            await vm.RefreshAsync(() => Csla.DataPortal.FetchAsync<Dashboard>());
            user = Csla.ApplicationContext.User;
        }
        catch (Exception ex)
        {
            ex;
        }
    }
}
```

It's seems the  ApplicationContext have been lost after that the page is refreshed (OnParametersSetAsync)

**During the investigation, i have notice that the very same application when hosting in Kestrel make the issue disappear.** 

So I have also give a try to publish the application on IIS with the in-process and the out-of-process hosting => the issue is back.

Last point, I have tried differents version of csla / projectTracker on the rebel PC :
- csla-5.1.0 : no issue
- csla-5.2.0 : issue detected
- csla-5.4.1 : issue detected
- csla-5.4.2 : issue detected

Does someone has already encounter this issue ?

Thanks a lot for you feedback in advance,

Best regards,
Cedric

**Comment by rockfordlhotka** [comment, marked as answer]
There are known issues with server-side Blazor in CSLA 5. Fixing them has been a massive effort, and includes many breaking changes - all of which are in CSLA 6.

**Reply by sylum** [reply]
Thx a lot for for tips. I looking forward to work with the new version 6. 
This gives me the opportunity to thank you and the csla team for this great framework.
Best regards,
Cedric
